### ISSUE TITLE
WinFSP breaking change

### ISSUE THREAD

**Issue by bitlogik** [Scille/parsec-cloud#1230]
Following the [WinFSP update from 1.7B1 to 1.7](https://github.com/Scille/parsec-cloud/commit/f0071fa007563818eaa442795fd533902921fe27), there's a small breaking change which appeared to be missed. This deals with the way WinFSP behaves when a non-existing directory is provided. Sadly, the unreliable testing pipeline about mountpoints in Windows hid this issue.
There's an [opened issue in winfspy repository](https://github.com/Scille/winfspy/issues/24), and can have some implications in this ParsecCloud software, in case there are some parts which are handling this error, which is now a different one. This issue opened is to make sure this modification is taken care in this code base also.

For now, we have identified the following potential points (not exhaustive) :
https://github.com/Scille/parsec-cloud/blob/367ef122b7f52c6034f563f45f4005c5ebd0a15a/parsec/core/gui/files_widget.py#L744
https://github.com/Scille/parsec-cloud/blob/af2ab5f886e0db851ef245f869cadd65b55c90d2/parsec/test_utils/workspacefs.py#L37

One way to fix this, can also be that winfspy catches this case, and throws STATUS_OBJECT_NAME_NOT_FOUND as before. This way, there won't be any impact in this Parsec code base.

**Comment by vxgmichel** [comment]
Hi @bitlogik,

I don't think this issue is related to the upgrade to winfsp 1.7, we had the same consistency issue with winfstest before the corresponding PR was merged. I'll keep investigating, thanks!

**Comment by bitlogik** [comment]
My POV is that the consistency issue with winfstest was hiding this issue. So we have now ```Expecting: ERROR_PATH_NOT_FOUND
-> Got: errno=267, message='The directory name is invalid.'``` as a recurring error in the Windows mountpoint tests.

**Comment by bitlogik** [comment]
Testing [from master with the following changes](https://github.com/bitlogik/parsec-cloud/commit/aef878dfe97be24dafdfbf0cd7f9daf184bde30f) :
- time sleep added
- remove xfail

```
tests/core/mountpoint/test_winfstest.py::test_winfstest[00_CreateFile_Dispositions.t] PASSED [ 80%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[01_CreateFile_Attributes.t] XFAIL [ 82%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[02_CreateRemoveDirectory.t] XFAIL [ 84%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[03_FindFiles.t] PASSED [ 86%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[04_MoveFile.t] XFAIL [ 88%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[05_CreateFile_Truncation.t] PASSED [ 90%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[06_FileTimes.t] XFAIL [ 92%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[07_SetGetFileAttributes.t] XFAIL [ 94%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[08_CreateFile_Delete.t] PASSED [ 96%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[09_CreateFile_Sharing.t] PASSED [ 98%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[10_GetSetSecurity.t] XFAIL [100%]
```

Now testing with a [revert of the WinFSP update](https://github.com/bitlogik/parsec-cloud/commit/00938afd9a699912ca83f449abe3b0b558c407d1) :
```
tests/core/mountpoint/test_winfstest.py::test_winfstest[00_CreateFile_Dispositions.t] PASSED [ 80%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[01_CreateFile_Attributes.t] XFAIL [ 82%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[02_CreateRemoveDirectory.t] XFAIL [ 84%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[03_FindFiles.t] PASSED [ 86%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[04_MoveFile.t] XFAIL [ 88%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[05_CreateFile_Truncation.t] PASSED [ 90%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[06_FileTimes.t] XFAIL [ 92%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[07_SetGetFileAttributes.t] XFAIL [ 94%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[08_CreateFile_Delete.t] PASSED [ 96%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[09_CreateFile_Sharing.t] PASSED [ 98%]
tests/core/mountpoint/test_winfstest.py::test_winfstest[10_GetSetSecurity.t] XFAIL [100%]
```

The exact same tests failed.
Was this issue about bad error catch in winfspy was present even before the update?

**Comment by vxgmichel** [comment]
Hi @bitlogik,

Actually, `XFAIL` means the test should be ignored:
https://github.com/Scille/parsec-cloud/blob/335bdb699684f593454320ba53d63f7501ea95c0/tests/core/mountpoint/test_winfstest.py#L34-L56

We use this to mark tests that fails but that we'd like to support in the future.

**Comment by bitlogik** [comment]
OK, we're not expert in pytest.
We're not sure about the root cause of the issue. Still, we're sure that this issue about error code in winfspy must be investigated further.

**Comment by vxgmichel** [comment]
> We're not sure about the root cause of the issue. Still, we're sure that this issue about error code in winfspy must be investigated further.

Yes, plus I've been able to observe the issue you mentioned earlier (`test_folder_operations` hanging the CI for one hour). I'll keep investigating it.

**Comment by vxgmichel** [comment]
Hi @bitlogik ,

The CI should be more stable now so the question is, are those consistency issues simply related to the azure CI environment or does it hide some deeper problems with winfsp? 

You seem to think that the WinFSP update introduced a breaking change but I'm not convinced of it. Would you care to explain what makes you think that?

I read the issue https://github.com/Scille/winfspy/issues/24, but I think you got confused with those error codes:
> From what we can understand, there's no more error thrown "ERROR_PATH_NOT_FOUND" (STATUS_OBJECT_NAME_NOT_FOUND = ENOENT), but the Win32 throw NotADirectoryError (WinError 267 = 'The directory name is invalid.')

- `STATUS_OBJECT_NAME_NOT_FOUND` is the [NTSTATUS](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55) that should be returned by a WinFSP operation when a given path does not exist
- `ERROR_PATH_NOT_FOUND` (error 3) is a [system error code](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-) return by the windows API calls when a program tries to access a file or directory that doesn't exist 
- `ERROR_DIRECTORY` (error 267) is also a [system error code](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-) return by the windows API calls when a program tries to perform a directory operation on an object that is not a directory

That being said, let's have a look at the failing statement:

https://github.com/Scille/winfspy/blob/01ea85541afcc7efb6c1e8494f1877a301dbad66/src/winfspy/tests/winfstest/cases/00_CreateFile_Dispositions.t#L49-L52

Here we try to open a file called `bar` in a directory that has the same name as a file that has just been deleted. Since neither a file or directory with this name exists, we expect a `ERROR_PATH_NOT_FOUND`. This is what happens most of time since this error does not happen too often.

But sometimes, it looks like the file is not correctly deleted (either because the deletion failed or because the deletion is still pending by the time we run the problematic statement). Then we get the `ERROR_DIRECTORY` error we discussed earlier.

In the past, we had some issues with deletion not being performed as soon as the windows call returned so there's a good chance those problems are related. In particular, it might be the same issue as the inconsistency in `test_mountpoint_revoke_access[write]`.

Since this issue seems to have more to do with the tests than the actual code, I'll close those issues. I'll create another one later dedicated to this particular issue. 

Thanks for the report :)

**Comment by bitlogik** [comment]
Sorry to point to a false direction. Indeed, we were very confused by the different level of errors, and also the fact the tests is about a directory with the same name as a deleted file. So the ERROR_DIRECTORY error means in this case the file was not deleted, and if deleted successfully ERROR_PATH_NOT_FOUND. I didn't fully understand these tests case, and their root cause.
The error "Expecting: ERROR_PATH_NOT_FOUND -> Got: errno=267" was so recurring since we updated WinFSP, I imagined there was an error in the tests script, not in the way they were run, they actually detected a really issue. The testing are very wide, Windows mountpoint tests are using "flaky repeated" behaviour, and and there bring some complexity in this project. We just expect we were helpful in helping you to investigate and dig about this matter.

So the root cause is the deletion command call returns without the file was properly deleted. This issue was not directly addressed so far. In the pipeline to fix the mountpoint tests pipeline, it only runs the tests until everything pass. Again, we're not sure to read this correctly, can you confirm this?
Can we open an issue in winfspy to address this issue, or at least keep it in mind : "file deletion callback thrown when sometimes not deleted" or a similar description ?
Maybe a basic small time delay after the deletion command can provide a good workaround ?
What is the impact of this issue? The application is never trying to reach a deleted file, after the delete command? So the impact is null on a standard use of the system? Is this deleted command only a command in the testing script, only for testing purpose? In this case a 100ms delay after file deletion command can do the trick.

**Comment by vxgmichel** [comment]
> In the pipeline to fix the mountpoint tests pipeline, it only runs the tests until everything pass. Again, we're not sure to read this correctly, can you confirm this?

More precisely, 4 mountpoint tests are marked to be re-run if one fails the first time. If the test fails a second time, it's considered a failure. Otherwise it's considered passed.

> Can we open an issue in winfspy to address this issue, or at least keep it in mind : "file deletion callback thrown when sometimes not deleted" or a similar description ?

I'd rather keep those issues on parsec-cloud until we found the root cause of the problem. I'll create one and tag it here.

> Maybe a basic small time delay after the deletion command can provide a good workaround ?

Maybe, but more investigation is necessary first.

> What is the impact of this issue? The application is never trying to reach a deleted file, after the delete command? So the impact is null on a standard use of the system? 

That would be my view yes. This might simply be how deletion work on windows, although we have to make sure of it

> Is this deleted command only a command in the testing script, only for testing purpose? In this case a 100ms delay after file deletion command can do the trick.

Yes, if the above hypothesis is correct we need to find a way to wait for those deletions to complete and fix the tests accordingly.

**Comment by vxgmichel** [comment]
Hi @bitlogik ,

I can confirm that this issue is fixed by #1235, it'll be merged soon :)
